File: Courses/3.Quera_advanced_algorithm_and_data_structure_course/9.Data_Structures/2.history.py
q = int(input())
curr_str = ""
history = []
for c in range(q):
    command = input().split()
    if(command[0] != "undo"):
        ind = int(command[1])-1
        history.append(curr_str)
        
    match command[0]:
        case "insert":
            curr_str = curr_str[:ind] + command[2] + curr_str[ind:]
        case "delete":
            curr_str = curr_str[:ind] + curr_str[ind+1:]
        case "undo":
            curr_str = history.pop()
            
    
print(curr_str)

File: 3.Quera_advanced_algorithm_and_data_structure_course/9.Data_Structures/2.history.cpp
#include <iostream>
#include <stack>
using namespace std;

int main(){
    int q; cin >> q;

    stack<string> history;
    history.push("");
    for(int i=0; i<q; i++){
        string command; cin >> command;

        string last_string = history.top();
        if(command == "insert"){
            int ind; cin >> ind; ind--;
            char ch; cin >> ch;
            for(int j=ind; j<last_string.size(); j++)
                swap(last_string[j], ch);
            last_string+=ch;

            history.push(last_string);
        }else if(command == "delete"){
            int ind; cin >> ind; ind--;

            string tmp_str = "";
            for(int j=0; j< last_string.size(); j++)
                if(j!=ind)
                    tmp_str += last_string[j];

            history.push(last_string);
        }else{
            history.pop();
        }

    }
    cout << history.top() << endl;


    return 0;
}